File: app/src/main/java/app/opass/ccip/network/ConfClient.java
package app.opass.ccip.network;

import java.util.List;

import app.opass.ccip.model.Submission;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;

public class ConfClient {

    public static final String API_BASE_URL = "https://summit.g0v.tw";

    private static Retrofit retrofit;
    private static ConfService sConfService;

    public static Retrofit getRetrofit() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(API_BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }

        return retrofit;
    }

    public static ConfService get() {
        if (sConfService == null) {
            sConfService = getRetrofit().create(ConfService.class);
        }

        return sConfService;
    }

    public interface ConfService {
        @GET("/2018/static/ccip.json")
        Call<List<Submission>> submission();
    }
}

File: app/src/main/java/app/opass/ccip/network/ConfClient.kt
package app.opass.ccip.network

import app.opass.ccip.model.Session
import app.opass.ccip.util.JsonUtil
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Url

class ConfClient {
    companion object {
        private const val API_BASE_URL = "https://opass.app"

        private val retrofit: Retrofit by lazy {
            Retrofit.Builder()
                .baseUrl(API_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(JsonUtil.GSON))
                .build()
        }
        private val sConfService: ConfService by lazy {
            retrofit.create(ConfService::class.java)
        }

        fun get() = sConfService

        interface ConfService {
            @GET
            fun session(@Url url: String): Call<List<Session>>
        }
    }
}
